src/action.ml

```diff
@@ -129,59 +135,83 @@ struct
            Diff { optional = false; file1; file2; mode = Binary })
         ])
 
-  let rec sexp_of_t : _ -> Sexp.t =
-    let path = Path.sexp_of_t and string = String.sexp_of_t in
-    function
-    | Run (a, xs) -> List (Sexp.unsafe_atom_of_string "run"
-                           :: Program.sexp_of_t a :: List.map xs ~f:string)
-    | Chdir (a, r) -> List [Sexp.unsafe_atom_of_string "chdir" ;
-                            path a ; sexp_of_t r]
-    | Setenv (k, v, r) -> List [Sexp.unsafe_atom_of_string "setenv" ;
-                                string k ; string v ; sexp_of_t r]
-    | Redirect (outputs, fn, r) ->
-      List [ Sexp.atom (sprintf "with-%s-to" (Outputs.to_string outputs))
-           ; path fn
-           ; sexp_of_t r
-           ]
-    | Ignore (outputs, r) ->
-      List [ Sexp.atom (sprintf "ignore-%s" (Outputs.to_string outputs))
-           ; sexp_of_t r
-           ]
-    | Progn l -> List (Sexp.unsafe_atom_of_string "progn"
-                       :: List.map l ~f:sexp_of_t)
-    | Echo xs ->
-      List (Sexp.unsafe_atom_of_string "echo" :: List.map xs ~f:string)
-    | Cat x -> List [Sexp.unsafe_atom_of_string "cat"; path x]
-    | Copy (x, y) ->
-      List [Sexp.unsafe_atom_of_string "copy"; path x; path y]
-    | Symlink (x, y) ->
-      List [Sexp.unsafe_atom_of_string "symlink"; path x; path y]
-    | Copy_and_add_line_directive (x, y) ->
-      List [Sexp.unsafe_atom_of_string "copy#"; path x; path y]
-    | System x -> List [Sexp.unsafe_atom_of_string "system"; string x]
-    | Bash   x -> List [Sexp.unsafe_atom_of_string "bash"; string x]
-    | Write_file (x, y) -> List [Sexp.unsafe_atom_of_string "write-file";
-                                 path x; string y]
-    | Rename (x, y) -> List [Sexp.unsafe_atom_of_string "rename";
-                             path x; path y]
-    | Remove_tree x -> List [Sexp.unsafe_atom_of_string "remove-tree"; path x]
-    | Mkdir x       -> List [Sexp.unsafe_atom_of_string "mkdir"; path x]
-    | Digest_files paths -> List [Sexp.unsafe_atom_of_string "digest-files";
-                                  List (List.map paths ~f:path)]
-    | Diff { optional; file1; file2; mode = Binary} ->
-      assert (not optional);
-      List [Sexp.unsafe_atom_of_string "cmp"; path file1; path file2]
-    | Diff { optional = false; file1; file2; mode = _ } ->
-      List [Sexp.unsafe_atom_of_string "diff"; path file1; path file2]
-    | Diff { optional = true; file1; file2; mode = _ } ->
-      List [Sexp.unsafe_atom_of_string "diff?"; path file1; path file2]
-    | Merge_files_into (srcs, extras, target) ->
-      List
-        [ Sexp.unsafe_atom_of_string "merge-files-into"
-        ; List (List.map ~f:path srcs)
-        ; List (List.map ~f:string extras)
-        ; path target
-        ]
+  module Make_gen (P : sig
```

Comment:
Let's only have the sexp_of_t function here, and in For_shell we convert first to a Sexp.t and then to a Dsexp.t

Reply:
Actually, I don't think that will work as it will not generate the same sexp's when we serialize them. E.g. for the inline tests subsystem.

Reply:
I didn't realize we were generating the constructors for paths in installed dune files. This seems wrong actually, this files should use a similar syntax to dune files. I guess we can leave it as it then